lib/sidekiq/xaclient.rb

```diff
@@ -0,0 +1,52 @@
+require "sidekiq/client"
+
+begin
+  require "after_commit_everywhere"
+rescue LoadError
+  Sidekiq.logger.error("You need to add after_commit_everywhere to your Gemfile for this to work")
+  exit(-127)
+end
+
+module Sidekiq
+  module XAClient
+    include AfterCommitEverywhere
+
+    ##
+    # Control job push within ActiveRecord transactions. Jobs can specify an
+    # "xa" attribute to define the policy to use:
+    #   * true or "commit" means enqueue the job after committing the current transaction.
+    #   * "rollback" means enqueue this job only if the current transaction rolls back
+    #   * nil or false means enqueue the job immediately, Sidekiq's default behavior
+    def push(item)
+      # Sidekiq::Job does not merge sidekiq_options so we need to fallback
+      policy = item.fetch("xa") { |key|
+        kl = item["class"]
+        kl.respond_to?(:get_sidekiq_options) ? kl.get_sidekiq_options[key] : nil
+      }
+      if policy == "commit" || policy == true
+        after_commit { super }
+      elsif policy == "rollback"
+        after_rollback { super }
+      else # enqueue immediately
+        super
+      end
+    end
+
+    ##
+    # We don't provide transactionality for push_bulk because we don't want
+    # to hold potentially hundreds of thousands of job records in memory due to
+    # a long running enqueue process. TODO: wdyt?
```

Comment:
As long as it's clear in the docs I think that it's a good decision. As you reverted most of the changes from #5158 I understand `push_bulk` as a special case. Combining two special cases increases the complexity and may therefore lead to unexpected results like the memory usage that you mentioned.

Reply:
Yeah, I want to keep push_bulk as simple as possible with efficient semantics so I think this is the best approach.